### PR TITLE
Add some infrastructure for snapshot integration test.

### LINKED ISSUE #1849: Snapshot create/restore integ tests infrastructure

### PR DESCRIPTION
**Pull request by sandreim**
## Reason for This PR
Fixes #1849.

## Description of Changes
Adds the flexibility required to generate a test matrix for save/restore tests using multiple snapshot images and firecracker binaries.

This PR introduces a new way for spinning up microvms in integration tests. Firstly, it does not rely on the hidden logic of fixtures to create and configure the microvms. Secondly, it defines a decoupled view of test resources, machine configurations, microvm configuration and implements  simple APIs that manage these. And thirdly, it allows a single test function to be run multiple times in a different setup (machine, disks, kernels, snapshots).

Layers and components:
- Resources layer components: **ArtifactCollection**, **Artifact**, **ArtifactSet**, **MicrovmBuilder**
- Test layer components: **TestContext**, **TestMatrix**

The resources exposed by these APIs are stored separately(`ci-artifacts` folder in S3 bucket) from the other resources used in integration tests.

The snapshot_pause_resume.py test was updated and now runs pause/resume on these 3 distinct microvm configurations: [2 vcpu 256MB](https://s3.amazonaws.com/spec.ccfc.min/ci-artifacts/microvms/2vcpu_256mb.json), [2 vcpu 512MB](https://s3.amazonaws.com/spec.ccfc.min/ci-artifacts/microvms/2vcpu_512mb.json), [2 vcpu 1024MB](https://s3.amazonaws.com/spec.ccfc.min/ci-artifacts/microvms/2vcpu_1024mb.json) in 8 seconds.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: tests/integration_tests/functional/test_pause_resume.py [hunk 1]
```diff
@@ -1,65 +1,108 @@
...
+    artifacts = ArtifactCollection(test_images_s3_bucket())
+
+    microvm_artifacts = artifacts.microvms()
+    kernel_artifacts = artifacts.kernels()
+    disk_artifacts = artifacts.disks()
+
+    # Create a test matrix. Push logger and network as context variables.
+    test_matrix = TestMatrix(context={
+            'network_config': network_config,
+            'logger': logger
```
Comment (gc-plp) [inline tests/integration_tests/functional/test_pause_resume.py:98]: I think that you could give a name to the logger here. What do you think?
You can _name_ logger instances by passing a string to the `getLogger` call, in order to have all log entries marked with the given name.

Comment (sandreim) [reply tests/integration_tests/functional/test_pause_resume.py]: done.

File: tests/framework/artifacts.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+MICROVM_KERNEL_EXTENSION = ".bin"
+MICROVM_DISK_EXTENSION = ".ext4"
+
+
+class Artifact:
+    """A generic artifact manipulation class."""
+
+    bucket = None
+    key = None
+    local_folder = None
```
Comment (gc-plp) [inline tests/framework/artifacts.py:28]: If `local_folder` is left initialized to `None`, an immediate call to `local_dir()` after instantiation would yield `None` as path.

It would be good to move the variables inside the `__init__` function and perhaps check for uninitialized values for `local_folder`. If `local_dir()` and `local_path()` are not designed to be public, the functions should also be marked as private.

File: tests/framework/artifacts.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(self, bucket, key, type="misc"):
```
Comment (gc-plp) [inline tests/framework/artifacts.py:30]: Could the artifact `type` be part of an enum, so that we explicitly allow only a predefined set of types?

File: tests/framework/artifacts.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        """Initialize bucket, key and type."""
+        self.bucket = bucket
+        self.key = key
+        self.type = type
+
+    def name(self):
+        """Return the artifact name."""
+        return self.key.rsplit("/", 1)[1]
```
Comment (gc-plp) [inline tests/framework/artifacts.py:38]: Instead of `rsplit`, you could also use `Path(self.key).name`

File: tests/framework/artifacts.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            self.name()
+        )
+
+
+class DiskArtifact(Artifact):
+    """Specializes the generic artifact."""
+
+    def ssh_key(self):
+        """Return a ssh key artifact."""
+        key_file_path = self.key.rsplit(".", 1)[0] + '.id_rsa'
```
Comment (gc-plp) [inline tests/framework/artifacts.py:69]: Same proposal here, as above. You could use `Path` to make the purpose of statement clearer.

File: tests/framework/artifacts.py (hunk 3 above)
Comment (sandreim) [reply tests/framework/artifacts.py:30]: I think this is a good idea.

File: tests/framework/artifacts.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            ):
+                artifacts.append(artifact_class(bucket,
+                                                file.key,
+                                                type=artifact_type))
+        return artifacts
+
+    def microvms(self, keyword=None):
+        """Return all microvms artifacts for the current arch."""
+        return self._fetch_artifacts(
+            self.ARTIFACTS_ROOT,
```
Comment (gc-plp) [inline tests/framework/artifacts.py:118]: Instead of `self` you can also reference the variables by `ArtifactCollection.ARTIFACTS_ROOT`. Using it this way is clearer that you're referencing a class variable that is read only.

File: tests/framework/artifacts.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+from shutil import copyfile
+from typing import List
+from os import path
+from pathlib import Path
+
+import boto3
+import botocore.client
+
+
+MICROVM_CONFIG_EXTENSION = ".json"
```
Comment (gc-plp) [inline tests/framework/artifacts.py:18]: Does any other class except `ArtifactCollection` use these definitions? Would it make more sense to either move them under `ArtifactCollection` or move the variables defined in `ArtifactCollection` inside an enum?

File: tests/framework/artifacts.py (hunk 6 above)
Comment (sandreim) [reply tests/framework/artifacts.py:118]: self is shorter 😃

File: tests/framework/artifacts.py (hunk 2 above)
Comment (sandreim) [reply tests/framework/artifacts.py:28]: Done

File: tests/framework/artifacts.py (hunk 6 above)
Comment (sandreim) [reply tests/framework/artifacts.py]: done

File: tests/framework/artifacts.py (hunk 5 above)
Comment (sandreim) [reply tests/framework/artifacts.py]: done

File: tests/integration_tests/functional/test_pause_resume.py [hunk 8]
```diff
@@ -1,65 +1,111 @@
...
     exit_code, _, _ = ssh_connection.execute_command("ls")
     assert exit_code == 0
+
+    microvm.kill()
+    del tap
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
```
Comment (ioanachirca) [inline tests/integration_tests/functional/test_pause_resume.py:80]: Pause/resume is available on ARM also.

File: tests/framework/artifacts.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+class ArtifactSet:
+    """Manages a set of artifacts with the same type."""
+
+    def __init__(self, artifacts):
+        """Initialize type and artifact array."""
+        assert len(artifacts) > 0
+        self._type = artifacts[0].type
+        self._artifacts = []
+        self.add(artifacts)
```
Comment (lauralt) [inline tests/framework/artifacts.py:169]: IMO it seems a little bit weird to have an `add` function called in an initialization function. If `add()` is only used here can we give up on this function and insert here, those artifacts (this way the second `assert` is not required)?
If we need in the future a function that adds new artifacts to the set we can make a separate one that will be used for this scope, i.e. adding (appending) not initializing :-??

File: tests/framework/artifacts.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+    def add(self, artifacts):
+        """Add artifacts to set."""
+        assert len(artifacts) > 0
+        for artifact in artifacts:
+            assert artifact.type == self._type
+            self._artifacts.append(artifact)
+
+    def artifact(self, index):
```
Comment (lauralt) [inline tests/framework/artifacts.py:178]: Is this used?
Also, just a personal opinion: the code gets hard to follow when there are lot of `assert`s (in non test files). Is it possible to replace them with `if`s and return `None` or some kind of error if those conditions are not met / reduce the number of `assert`s / group them on the same line?

File: tests/integration_tests/functional/test_pause_resume.py (hunk 8 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_pause_resume.py:80]: They work indeed, but have no real usecase by themselves.

File: tests/framework/artifacts.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_DISKS,
+            ArtifactCollection.MICROVM_DISK_EXTENSION,
+            ArtifactType.DISK,
+            DiskArtifact,
+            keyword=keyword
+        )
+
+
+class ArtifactSet:
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/artifacts.py:161]: In this context `Set` does not relate any language but to a concept: a collection of artifacts that have the same type similar to the noun `set`. The ArtifactSet class offers the guarantee that all artifacts have the same type.

File: tests/framework/artifacts.py (hunk 9 above)
Comment (sandreim) [reply tests/framework/artifacts.py:169]: `add()` contains logic that guarantees all artifacts in set are the same. It can be called to add more artifacts to the set after construction. Why would I duplicate it ?

File: tests/framework/artifacts.py (hunk 10 above)
Comment (sandreim) [reply tests/framework/artifacts.py:178]: I think you are right, this fn was used at some point but forgot to delete it as I was refactoring the code. 

Asserts are important to enforce guarantees so that undefined behavior does not propagate 
to actual test files.

File: tests/framework/builder.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+        # Download ssh key into microvm root.
+        ssh_key.download(self.root_path)
+        vm.ssh_config['ssh_key_path'] = ssh_key.local_path()
+        os.chmod(vm.ssh_config['ssh_key_path'], 0o400)
+
+        # Start firecracker.
+        vm.spawn()
+
+        # Apply the microvm artifact configuration and start microvm.
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/builder.py:48]: I'll move the comment before the `basic_config()` call.

File: tests/framework/artifacts.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+
+class Artifact:
+    """A generic artifact manipulation class."""
+
+    def __init__(self, bucket, key, artifact_type=ArtifactType.MISC):
+        """Initialize bucket, key and type."""
+        self.bucket = bucket
+        self.key = key
+        self.local_folder = None
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/artifacts.py:32]: Sure, makes sense.

File: tests/framework/artifacts.py (hunk 10 above)
Comment (gc-plp) [reply tests/framework/artifacts.py:178]: `assert`s are a good way of detecting whether an invalid state is detected from which the execution path will not recover.  

Since we don't always want to recover from a faulty test setup, I think that asserts are useful.

Perhaps explaining why they're there would be nice - pytest supports assert messages as shown [here](https://docs.pytest.org/en/latest/assert.html):

```python
assert a % 2 == 0, "value was odd, should be even"
```

File: tests/framework/matrix.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        self._set_index = 0
+        if not os.path.exists(cache_dir):
+            os.mkdir(cache_dir)
+
+    @property
+    def sets(self):
+        """Return the artifact sets."""
+        return self._sets
+
+    def add_set(self, artifact_set: ArtifactSet):
```
Comment (lauralt) [inline tests/framework/matrix.py:106]: This also looks unused. Can we add it when it will be used? It is easier to review this way :D.

File: tests/framework/artifacts.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+        self,
+        bucket
+    ):
+        """Initialize S3 client."""
+        config = botocore.client.Config(signature_version=botocore.UNSIGNED)
+        # pylint: disable=E1101
+        # fixes "E1101: Instance of '' has no 'Bucket' member (no-member)""
+        self.bucket = boto3.resource('s3', config=config).Bucket(bucket)
+
+    def _fetch_artifacts(self,
```
Comment (lauralt) [inline tests/framework/artifacts.py:105]: Do we plan on calling this with `artifact_root` different than `ArtifactCollection.ARTIFACTS_ROOT`?

File: tests/framework/artifacts.py (hunk 9 above)
Comment (sandreim) [reply tests/framework/artifacts.py:169]: Renamed to insert().

File: tests/framework/builder.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        fc_binary, jailer_binary = build_tools.get_firecracker_binaries()
+
+        vm = Microvm(
+            resource_path=self.root_path,
+            fc_binary_path=fc_binary,
+            jailer_binary_path=jailer_binary,
+            microvm_id=microvm_id,
+            bin_cloner_path=self.bin_cloner_path
+        )
+        vm.setup()
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/builder.py:33]: Unfortunately it doesn't work 😢 if I try to import it from conftest.py . This function will need to be changed in the future when we add firecracker binary artifacts, so indeed we are duplicating some code, with the promise that it will diverge at some point soon.

File: tests/framework/builder.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+class MicrovmBuilder:
+    """Build fresh microvms or restore from snapshot."""
+
+    def __init__(self, root_path, bin_cloner_path):
+        """Initialize microvm root and cloning binary."""
+        self.root_path = root_path
+        self.bin_cloner_path = bin_cloner_path
+
+    def build(self, kernel: Artifact, disks: [DiskArtifact], config: Artifact):
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/builder.py:21]: This is required to build a microvm with multiple block devices, but there is not such test defined yet. There is more work pending here, like for example adding NetDeviceArtifact support.

File: tests/framework/builder.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        with open(config.local_path()) as microvm_config_file:
+            microvm_config = json.load(microvm_config_file)
+
+        vm.basic_config(vcpu_count=int(microvm_config['vcpu_count']),
+                        mem_size_mib=int(microvm_config['mem_size_mib']),
+                        ht_enabled=bool(microvm_config['ht_enabled']))
+
+        return vm
+
+    def build_from_snapshot(self, snapshot: Artifact):
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/builder.py:58]: I'll post a PR with with snapshots this week or next week with such a test. I'll add a comment here explaining how this is going to work.

File: tests/integration_tests/functional/test_pause_resume.py [hunk 19]
```diff
@@ -1,65 +1,111 @@
...
-    test_microvm = test_microvm_with_ssh
-    test_microvm.spawn()
-
-    # Set up the microVM with 2 vCPUs, 256 MiB of RAM and a root file
-    # system with the rw permission.
-    test_microvm.basic_config()
-
-    # Create tap before configuring interface.
-    _tap1, _, _ = test_microvm.ssh_network_config(network_config, '1')
+"""Basic tests scenarios for snapshot save/restore."""
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/integration_tests/functional/test_pause_resume.py:3]: done

File: tests/framework/matrix.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+"""Generate multiple microvm configurations and run tests.
+
+Implements a component that calls a specific function in the context
+of the cartesian product of all artifact sets.
+"""
+
+import os
+from framework.artifacts import ArtifactSet
+
+
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/matrix.py:12]: done

File: tests/framework/artifacts.py [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+class ArtifactCollection:
+    """Provides easy access to different artifact types."""
+
+    MICROVM_CONFIG_EXTENSION = ".json"
+    MICROVM_KERNEL_EXTENSION = ".bin"
+    MICROVM_DISK_EXTENSION = ".ext4"
+
+    # S3 bucket structure.
+    ARTIFACTS_ROOT = 'ci-artifacts'
+    ARTIFACTS_DISKS = '/disks/' + platform.machine()
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/artifacts.py:90]: done.

File: tests/integration_tests/functional/test_pause_resume.py [hunk 22]
```diff
@@ -1,65 +1,111 @@
...
+
+    logger.info("Testing microvm: \"{}\" with kernel {} and disk {} "
+                .format(context.microvm.name(),
+                        context.kernel.name(),
+                        context.disk.name()))
+
+    microvm = vm_builder.build(context.kernel,
+                               [context.disk],
+                               context.microvm)
+    tap = microvm.ssh_network_config(context.custom['network_config'], '1')
```
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/integration_tests/functional/test_pause_resume.py:26]: It is used, please see below del tap. If I do not explicitly delete it I get these:
```
Cannot open network namespace "82b35ec3-1ffb-4a45-b214-b8aafa52af45": No such file or directory

Returned error code: 1
stderr:
Cannot open network namespace "82b35ec3-1ffb-4a45-b214-b8aafa52af45": No such file or directory
```

File: tests/framework/matrix.py (hunk 14 above)
Comment (sandreim) [reply tests/framework/matrix.py:106]: Removed it, the constructor is enough.

File: tests/framework/artifacts.py (hunk 15 above)
Comment (sandreim) [reply tests/framework/artifacts.py:105]: Nope, I've removed that parameter.

**Comment by sandreim** [comment]
@lauralt @ioanachirca PTAL.

File: tests/framework/builder.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/framework/builder.py:21]: Okay, I thought this function is designed to do just a basic configuration and adding non root block devices, net devices, etc. will be done separately.

File: tests/framework/builder.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/framework/builder.py:33]: Hmm, I just tried that now and it worked for me :-?.
```python
from conftest import init_microvm
...
vm = init_microvm(self.root_path, self.bin_cloner_path)
```
Isn't this working or am I doing something wrong?

File: tests/integration_tests/functional/test_pause_resume.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/integration_tests/functional/test_pause_resume.py:27]: Yep, `del _tap` is needed indeed, but it looks like if we only use that variable for deleting it later, python considers `tap` as unused and marks it with that gray annoying highlight, so that's why I've suggested to rename it to `_tap`.

File: tests/integration_tests/functional/test_pause_resume.py [hunk 23]
```diff
@@ -1,65 +1,112 @@
...
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_pause_resume(test_session_root_path,
+                      network_config,
+                      bin_cloner_path):
+    """Test scenario: boot/pause/resume for all available configurations."""
+    logger = logging.getLogger("pause_resume")
+    # Currently, artifacts share the bucket with all other resources.
+    artifact_collection = ArtifactCollection(test_images_s3_bucket())
```
Comment (lauralt) [inline tests/integration_tests/functional/test_pause_resume.py:89]: Would be nice to add a comment here why we have that `integration_tests/functional/test_pause_resume.py::test_images_s3_bucket` in the `tools/devtool test` output (in case you didn't fix that meantime).

File: tests/framework/matrix.py [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+    """Computes the cartesian product of artifacts."""
+
+    __test__ = False
+
+    def __init__(self,
+                 artifact_sets,
+                 context=TestContext(),
+                 cache_dir=ARTIFACTS_LOCAL_ROOT):
+        """Initialize the cache directory."""
+        self._cache_dir = cache_dir
```
Comment (lauralt) [inline tests/framework/matrix.py:88]: nit: Would it make sense to move this at the end of the function? It makes things a little bit clearer I think to initialize that field after being sure that `cache_dir` path exists.

File: tests/framework/matrix.py [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+
+            # Pop the previous artifact from solution.
+            cartesian_product.pop()
+            # Walk down 1 level.
+            self._set_index -= 1
+
+    def _run_test_fn(self, artifacts, test_fn):
+        """Patch context and call test_fn."""
+        for artifact in artifacts:
+            self._context.set_any(artifact.type.value, artifact)
```
Comment (lauralt) [inline tests/framework/matrix.py:136]: nit: it would be nice to have a function in `Artifact` class that returns `artifact.type.value`.

Comment (sandreim) [reply tests/framework/matrix.py:136]: The type is currently an accessor for the type (which is an enum). Value is an accesor for the Enum object.

File: tests/framework/matrix.py (hunk 24 above)
Comment (sandreim) [reply tests/framework/matrix.py:88]: The cache dir needs to be initialized regardless if it exists.

Comment (lauralt) [reply tests/framework/matrix.py:88]:
```python
        self._context = context
        # Stores the artifact sets array.
        self._sets = artifact_sets
        # ArtifactSet stack pointer.
        self._set_index = 0
        if not os.path.exists(cache_dir):
            os.mkdir(cache_dir)
        self._cache_dir = cache_dir
```

That's what I suggested, it seems somehow more natural this flow, but it's a nit anyway.

File: tests/framework/matrix.py (hunk 25 above)
Comment (lauralt) [reply tests/framework/matrix.py:136]: Yep, I preferred to kind of hide at this level the `type` type (enum), and just call a function from `Artifact` class: `artifact.type_value()`
```
def type_value(self):
        """Return the artifact type value."""
        return self._type.value
```
(but that's pretty subjective I think :-?)

File: tests/framework/builder.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sandreim) [reply tests/framework/builder.py:58]: Added comment.

File: tests/integration_tests/functional/test_pause_resume.py (hunk 23 above)
Comment (sandreim) [reply tests/integration_tests/functional/test_pause_resume.py:89]: I've reverted back to `from conftest import init_microvm`. Seems to work now, maybe there was something wrong locally.

The PR does not fail on unused checks, neither is my IDE. It is used as it is reference by the delete.

File: tests/framework/matrix.py (hunk 24 above)
Comment (sandreim) [reply tests/framework/matrix.py:88]: done.

File: tests/framework/matrix.py (hunk 25 above)
Comment (sandreim) [reply tests/framework/matrix.py:136]: We can just use strings instead of enums if we make that change. Artifact types are pretty much opaque to the tests as they are fetched and manipulated through proper APIs.

File: tests/framework/builder.py [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+class MicrovmBuilder:
+    """Build fresh microvms or restore from snapshot."""
+
+    def __init__(self, root_path, bin_cloner_path):
+        """Initialize microvm root and cloning binary."""
+        self.root_path = root_path
+        self.bin_cloner_path = bin_cloner_path
+
+    def build(self, kernel: Artifact, disks: [DiskArtifact], config: Artifact):
+        """Build and start a fresh microvm."""
```
Comment (ioanachirca) [inline tests/framework/builder.py:23]: This comment might have slipped through - the method only builds the microvm.

File: tests/framework/builder.py [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+        vm.basic_config(vcpu_count=int(microvm_config['vcpu_count']),
+                        mem_size_mib=int(microvm_config['mem_size_mib']),
+                        ht_enabled=bool(microvm_config['ht_enabled']))
+
+        return vm
+
+    # TBD: Snapshot support is not fully merged. Once that is read
+    # this function will provide a way to spin up clones.
+    def build_from_snapshot(self, snapshot: Artifact):
+        """Build and start a microvm from a snapshot artifact."""
```
Comment (ioanachirca) [inline tests/framework/builder.py:54]: The method above only builds the microvm, will it be the same for this one?

File: tests/framework/builder.py (hunk 26 above)
Comment (sandreim) [reply tests/framework/builder.py:23]: It also starts it in basic_config().

File: tests/framework/builder.py (hunk 27 above)
Comment (sandreim) [reply tests/framework/builder.py:53]: Not sure yet, it's either paused or running.

File: tests/conftest.py [hunk 28]
```diff
@@ -117,15 +117,15 @@ def test_with_any_microvm(test_microvm_any):
     raise PermissionError("Test session needs to be run as root.")
 
 
-def test_images_s3_bucket():
+def _test_images_s3_bucket():
```
Comment (lauralt) [inline tests/conftest.py:120]: :tada:
